library/include/rocblas-types.h

```diff
@@ -38,9 +38,18 @@ typedef int64_t rocblas_stride;
 #endif
 
 // floating point types
-typedef float    rocblas_float;
-typedef double   rocblas_double;
-typedef uint16_t rocblas_half; // TODO: should be replaced with a struct, to become a unique type
+typedef float  rocblas_float;
+typedef double rocblas_double;
+
+// Clang supports _Float16 as an extension in C11 and C++11 and later
+#if __cplusplus >= 201103L || __STDC_VERSION__ >= 201112L
+typedef _Float16 rocblas_half;
+#else
+typedef struct
+{
+    uint16_t data;
+} rocblas_half;
+#endif
```

Comment:
> typedef struct
> {
 >   uint16_t data;
> } rocblas_half;
> #endif
>
> Will this case even be handled by the code now?

That is needed when someone compiles with a C/C++ compiler which doesn't recognize `_Float16` but only needs to pass pointers to 16-bit values to rocBLAS. It is similar to the old `rocblas_half`, except that it has a wrapper `struct` to distinguish it from `uint16_t`, which, under the old scheme, could not be disambiguated from `rocblas_half` since `rocblas_half` was simply a `typedef uint16_t rocblas_half;`

Clang supports `_Float16` in C11 and C++11 and later, hence the version checks around this declaration. GCC has wavered, sometimes including it in C++ mode but sometimes saying it shouldn't be, because it's only mentioned in the C11 standard and not in C++11.  There is no feature flag that I know of for detecting `_Float16` support, which isn't necessarily defined as a macro, although maybe with SFINAE tricks you could figure it out in C++ (something we cannot do in this header file, which must compile in C as well as C++).

If we need to add GCC-specific checks for `_Float16` support, they can be added.

We were already using `_Float16` when calling the old Tensile client; this just standardizes rocBLAS to use `_Float16` for `rocblas_half` and to avoid using any other type names in the code, so that we have one definition for the whole library.

Eventually, all C++ compilers will support `_Float16`, `__fp16`, or some variant.

Reply:
Okay, I just wasn't sure all your other changes to implicit conversions to float will still work for these old compilers that use this struct.

Reply:
> Okay, I just wasn't sure all your other changes to implicit conversions to float will still work for these old compilers that use this struct.

We won't be using the old compilers to build the code which does the conversions to `float`. We will be using compilers which support `_Float16`, which can be implicitly converted to `float` and `double`.

For old compilers, we just need an opaque datatype which can be `memcpy`'ed from FP16 data, and pointers to `rocblas_half` can be converted to/from pointers to FP16 data.

We may need to add more checks to handle GCC, since, unlike Clang, GCC has not expressed universal support for `_Float16` in C++.

If you are using or allowing GCC to be used for the `examples`, then we might have to account for it, and make sure that they build and run correctly.